txnkv/transaction/pipelined_flush.go

```diff
@@ -0,0 +1,448 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transaction
+
+import (
+	"bytes"
+	"github.com/docker/go-units"
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/kvrpcpb"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/tikv/client-go/v2/config/retry"
+	tikverr "github.com/tikv/client-go/v2/error"
+	"github.com/tikv/client-go/v2/internal/client"
+	"github.com/tikv/client-go/v2/internal/locate"
+	"github.com/tikv/client-go/v2/internal/logutil"
+	"github.com/tikv/client-go/v2/tikvrpc"
+	"github.com/tikv/client-go/v2/txnkv/txnlock"
+	"github.com/tikv/client-go/v2/util"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+type actionPipelinedFlush struct {
+	generation uint64
+}
+
+var _ twoPhaseCommitAction = actionPipelinedFlush{}
+
+func (action actionPipelinedFlush) String() string {
+	return "pipelined_flush"
+}
+
+func (action actionPipelinedFlush) tiKVTxnRegionsNumHistogram() prometheus.Observer {
+	return nil
+}
+
+func (c *twoPhaseCommitter) buildPipelinedFlushRequest(batch batchMutations, generation uint64) *tikvrpc.Request {
+	m := batch.mutations
+	mutations := make([]*kvrpcpb.Mutation, m.Len())
+
+	for i := 0; i < m.Len(); i++ {
+		assertion := kvrpcpb.Assertion_None
+		if m.IsAssertExists(i) {
+			assertion = kvrpcpb.Assertion_Exist
+		}
+		if m.IsAssertNotExist(i) {
+			assertion = kvrpcpb.Assertion_NotExist
+		}
+		mutations[i] = &kvrpcpb.Mutation{
+			Op:        m.GetOp(i),
+			Key:       m.GetKey(i),
+			Value:     m.GetValue(i),
+			Assertion: assertion,
+		}
+	}
+
+	minCommitTS := c.startTS + 1
+
+	req := &kvrpcpb.FlushRequest{
+		Mutations:   mutations,
+		PrimaryKey:  c.primary(),
+		StartTs:     c.startTS,
+		MinCommitTs: minCommitTS,
+		Generation:  generation,
+	}
+
+	r := tikvrpc.NewRequest(
+		tikvrpc.CmdFlush, req, kvrpcpb.Context{
+			Priority:               c.priority,
+			SyncLog:                c.syncLog,
+			ResourceGroupTag:       c.resourceGroupTag,
+			DiskFullOpt:            c.diskFullOpt,
+			TxnSource:              c.txnSource,
+			MaxExecutionDurationMs: uint64(client.MaxWriteExecutionTime.Milliseconds()),
+			RequestSource:          c.txn.GetRequestSource(),
+			ResourceControlContext: &kvrpcpb.ResourceControlContext{
+				ResourceGroupName: c.resourceGroupName,
+			},
+		},
+	)
+	return r
+}
+
+func (action actionPipelinedFlush) handleSingleBatch(
```

Comment:
It should not block this PR but there seems to be many unnecessary duplicated code for various actions...